---
title: "MICB 405 Project 2"
author: "Shannah"
date: "November 17, 2018"
output:
  word_document: default
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE, warning=FALSE)
library(tidyr)
library(dplyr)
library(pathview)
library(tidyverse)
library(knitr)
library(ggthemes)
library(grid)

```

```{r}
#Before we get started, I am going to load a function to create a theme from https://rpubs.com/Koundy/71792, which is a nice theme that I will use throughout the rest of the plots.

theme_Publication <- function(base_size=12) {
      (theme_foundation(base_size=base_size)
       + theme(plot.title = element_text(face = "bold",
                                         size = rel(1.2), hjust = 0.5),
               text = element_text(),
               panel.background = element_rect(colour = NA),
               plot.background = element_rect(colour = NA),
               panel.border = element_rect(colour = NA),
               axis.title = element_text(face = "bold"), #,size = rel(1)
               axis.title.y = element_text(angle=90,vjust =2),
               axis.title.x = element_text(vjust = -0.2),
               axis.text = element_text(), 
               axis.line = element_line(colour="black"),
               axis.ticks = element_line(),
               panel.grid.major = element_line(colour="#f0f0f0"),
               panel.grid.minor = element_blank(),
               legend.key = element_rect(colour = NA),
               legend.position = "bottom",
               legend.direction = "horizontal",
               legend.key.size= unit(0.2, "cm"),
               legend.margin = unit(0, "cm"),
               legend.title = element_text(face="italic"),
               plot.margin=unit(c(10,5,5,5),"mm"),
               strip.background=element_rect(colour="#f0f0f0",fill="#f0f0f0"),
               strip.text = element_text(face="bold")
          ))
      
}

theme_Publication_legend_side <- function(base_size=12) {
      (theme_foundation(base_size=base_size)
       + theme(plot.title = element_text(face = "bold",
                                         size = rel(1.2), hjust = 0.5),
               text = element_text(),
               panel.background = element_rect(colour = NA),
               plot.background = element_rect(colour = NA),
               panel.border = element_rect(colour = NA),
               axis.title = element_text(face = "bold"), #,size = rel(1)
               axis.title.y = element_text(angle=90,vjust =2),
               axis.title.x = element_text(vjust = -0.2),
               axis.text = element_text(), 
               axis.line = element_line(colour="black"),
               axis.ticks = element_line(),
               panel.grid.major = element_line(colour="#f0f0f0"),
               panel.grid.minor = element_blank(),
               legend.key = element_rect(colour = NA),
               legend.key.size= unit(0.2, "cm"),
               legend.margin = unit(0, "cm"),
               legend.title = element_text(face="italic"),
               plot.margin=unit(c(10,5,5,5),"mm"),
               strip.background=element_rect(colour="#f0f0f0",fill="#f0f0f0"),
               strip.text = element_text(face="bold")
          ))
      
}

theme_Publication_facets <- function(base_size=12) {
      (theme_foundation(base_size=base_size)
       + theme(plot.title = element_text(face = "bold",
                                         size = rel(1.2), hjust = 0.5),
               text = element_text(),
               panel.background = element_rect(colour = NA),
               plot.background = element_rect(colour = NA),
               axis.title = element_text(face = "bold"), #,size = rel(1)
               axis.title.y = element_text(angle=90,vjust =2),
               axis.title.x = element_text(vjust = -0.2),
               axis.text = element_text(), 
               axis.line = element_line(colour="black"),
               axis.ticks = element_line(),
               panel.grid.major = element_line(colour="#f0f0f0"),
               panel.grid.minor = element_blank(),
               legend.key = element_rect(colour = NA),
               legend.position = "bottom",
               legend.direction = "horizontal",
               legend.key.size= unit(0.2, "cm"),
               legend.margin = unit(0, "cm"),
               legend.title = element_text(face="italic"),
               plot.margin=unit(c(10,5,5,5),"mm"),
               strip.background=element_rect(colour="#f0f0f0",fill="#f0f0f0"),
               strip.text = element_text(face="bold")
          ))
      
}
```

# Data
Load the data and put it in a format that we need for plotting
```{r}

#--- GETTING THE DATA ----

########################
# ---- Saanich Data ----
# Load raw data
raw_dat <- readr::read_csv("Saanich_Data.csv")

#Clean dat
dat <- 
  raw_dat %>%
  dplyr::filter(!is.na(WS_O2)) %>%
  dplyr::rename(O2_uM=WS_O2, NO3_uM=WS_NO3, H2S_uM=WS_H2S) %>%
  dplyr::mutate(Depth_m=Depth*1000)

########################
# ---- checkM data ----
checkm_dat <- read.table("MetaBAT2_SaanichInlet_200m_min1500_checkM_stdout.tsv",
                         header=TRUE,
                         sep="\t",
                         comment.char = '') %>% 
  dplyr::rename(mag = Bin.Id) %>% 
  dplyr::select(mag, Completeness, Contamination)

# Due to a bug in the renaming script we have to rename the bins. Its a bit hacky but works using tidyverse functions
metag_rpkm <- read.table("SaanichInlet_200m_binned.rpkm.csv", header=T, sep=',') %>% 
  mutate(Sequence = gsub('m_', 'm.', Sequence)) %>% 
  mutate(Sequence = gsub('Inlet_', 'Inlet.', Sequence)) %>% 
  separate(col=Sequence, into=c("mag", "contig"), sep='_', extra="merge") %>% 
  group_by(Sample, mag) %>% 
  summarise(g_rpkm = mean(RPKM)) %>% 
  mutate(mag = gsub('Inlet.', 'Inlet_', mag)) 
  # dplyr::filter(Sample != "SI042_200m")

########################
# ---- rpkm and mag data ----
#Read in KO Annotation tables
ko <- read.table("SaanichInlet_MAGs_ORFs_ko.cleaned.txt") %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(ko = V2)

#Read in all rpkm files from each cruise in August
rpkm_48 <- read.table("SI048_200m.RPKM.csv", header=FALSE, sep=',') %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(rpkm = V2)
rpkm_72 <- read.table("SI048_200m.RPKM.csv", header=FALSE, sep=',') %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(rpkm = V2)
rpkm_73 <- read.table("SI073_200m.RPKM.csv", header=FALSE, sep=',') %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(rpkm = V2)
rpkm_74 <- read.table("SI074_200m.RPKM.csv", header=FALSE, sep=',') %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(rpkm = V2)
rpkm_75 <- read.table("SI075_200m.RPKM.csv", header=FALSE, sep=',') %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(rpkm = V2)

#Read in prokka MAG map
prokka_mag_map <- read.table("Prokka_MAG_map_basename.csv", header=F, sep=',') %>% 
  dplyr::rename(prokka_id = V1) %>% 
  dplyr::rename(mag = V2)

#Read in gtdbtk files
arc_class <- read.table("gtdbtk.ar122.classification_pplacer.tsv", sep="\t")
bac_class <- read.table("gtdbtk.bac120.classification_pplacer.tsv", sep="\t")

#Combine archaea and bacteria gtdb files
gtdb_dat <- rbind(arc_class, bac_class) %>% 
  dplyr::rename(mag = V1) %>% 
  separate(V2, sep=';', into=c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"))

#We can also determine the number of Class present in our bins, and how many MAGs are representing each.
gtdb_dat %>% 
  group_by(Order) %>% 
  summarise(count = n_distinct(mag)) %>% 
  knitr::kable()
gtdb_dat <- dplyr::select(gtdb_dat, mag, Kingdom, Phylum, Class, Order, Family)

#combine all cruises and ko values
ko_rpkm_48 <- left_join(ko, rpkm_48, by="orf") %>% 
  separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF
  left_join(prokka_mag_map, by="prokka_id") %>% 
  left_join(gtdb_dat, by="mag") %>%
  left_join(checkm_dat, by="mag") %>%
  left_join(metag_rpkm, by="mag")
ko_rpkm_72 <- left_join(ko, rpkm_72, by="orf") %>% 
  separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF
  left_join(prokka_mag_map, by="prokka_id") %>% 
  left_join(gtdb_dat, by="mag")%>%
  left_join(checkm_dat, by="mag") %>%
  left_join(metag_rpkm, by="mag")
ko_rpkm_73 <- left_join(ko, rpkm_73, by="orf") %>% 
  separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF
  left_join(prokka_mag_map, by="prokka_id") %>% 
  left_join(gtdb_dat, by="mag")%>%
  left_join(checkm_dat, by="mag") %>%
  left_join(metag_rpkm, by="mag")
ko_rpkm_74 <- left_join(ko, rpkm_74, by="orf") %>% 
  separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF
  left_join(prokka_mag_map, by="prokka_id") %>% 
  left_join(gtdb_dat, by="mag")%>%
  left_join(checkm_dat, by="mag") %>%
  left_join(metag_rpkm, by="mag")
ko_rpkm_75 <- left_join(ko, rpkm_75, by="orf") %>% 
  separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF
  left_join(prokka_mag_map, by="prokka_id") %>% 
  left_join(gtdb_dat, by="mag")%>%
  left_join(checkm_dat, by="mag") %>%
  left_join(metag_rpkm, by="mag")

#combine all ko_rpkm data into one data frame
rpkm_dat <- rbind(ko_rpkm_48, ko_rpkm_73, ko_rpkm_74, ko_rpkm_75)
head(rpkm_dat) %>% knitr::kable()

```

# Contamination vs Completeness Scatter Plot
First I will analyse the MAGs from checkM output with rpkm abundance values
Completeness and contamination ranges are based on quality standard definition from Bowers et al., 2017
```{r}

#TODO: include a None lineage?
#TODO: include range labels

rpkm_dat <- left_join(metag_rpkm, checkm_dat, by="mag") %>% 
  left_join(gtdb_dat, by="mag") %>% 
  group_by(mag, Kingdom, Phylum, Class, Completeness, Contamination) %>% 
  summarise(g_rpkm = mean(g_rpkm))

#-- PLOT WITH ALL OF THE DATA! --
#Plotting Completeness vs Contamination of MAGs at 200m
rpkm_dat %>%
  ggplot(aes(x=Completeness, y=Contamination, col=Class)) +
  # scale_size(range = c(1,10)) +
  # xlim(c(50,100)) +
  # ylim(c(0,100)) +
  geom_point() +
  #added discrete scales ; palette from colorbrewer2 but added extra ("#969696") b/c 13 values
  # scale_color_manual(values = c('#a6cee3','#1f78b4',"#b2df8a","#33a02c","#fb9a99",
                                # "#e31a1c",'#fdbf6f',"#ff7f00","#cab2d6","#6a3d9a",
                                # "#ffff99","#b15928", "#969696")) + 
  geom_vline(xintercept = 50, linetype = "dashed", size = 1, color = "#808080") +   #med completeness quality
  geom_vline(xintercept = 90, linetype = "dashed", size = 1, color = "#808080") +   #high completeness quality
  geom_hline(yintercept = 5, linetype = "dashed", size = 0.65, color = "#808080") +    #med contamination quality = 5-10%
  geom_hline(yintercept = 10, linetype = "dashed", size = 0.65, color = "#808080") +   #high contamination quality = > 10%
  xlab("Completeness (%)") +
  ylab("Contamination (%)") +
  theme_Publication_legend_side() 
  # guides(color=guide_legend(nrow=5)) 
    # labs(title="Completeness vs Contamination by Lineage")

#-- PLOT WITH CONTAMINATION <= 100%! --
#Plotting completeness vs contamination of MAGS at 200m, where contamination <=100%
# Note: contamination above 100% indicates that the recovered bin likely contains multiple organisms. To analyze closer, this plot removes the contamination that is > 100%.
rpkm_dat %>%
  dplyr::filter(Contamination < 100) %>%
  # dplyr::filter(Completeness > 50) %>%
  ggplot(aes(x=Completeness, y=Contamination, colour=Class)) +
  geom_point(aes(size=g_rpkm)) +
  #added discrete scales ; palette from colorbrewer2 but added extra ("#969696") b/c 13 values
  # scale_color_manual(values = c('#a6cee3','#1f78b4',"#b2df8a","#33a02c","#fb9a99",
                                # "#e31a1c",'#fdbf6f',"#ff7f00","#cab2d6","#6a3d9a",
                                # "#ffff99","#b15928", "#969696")) + 
  #-- LOW COMPLETENESS--
  annotate("text", x=25, y=100, label="Low", 
            colour="black", text=element_text(size=9, family="Arial")) +
  geom_segment(aes(x = 49, y = 96, xend = 1, yend = 96),
                  arrow = arrow(length = unit(0.25, "cm")), colour="black")+
  #-- MED COMPLETENESS--
  geom_vline(xintercept = 50, linetype = "dashed", size = 1, color = "#808080") +   #med completeness quality
  annotate("text", x=70, y=100, label="Medium", 
            colour="black", text=element_text(size=9, family="Arial")) +
  geom_segment(aes(x = 51, y = 96, xend = 89, yend = 96),
                  arrow = arrow(length = unit(0.25, "cm")), colour="black")+
  geom_segment(aes(x = 89, y = 96, xend = 51, yend = 96),
                  arrow = arrow(length = unit(0.25, "cm")), colour="black")+
  #-- HIGH COMPLETENESS --
  geom_vline(xintercept = 90, linetype = "dashed", size = 1, color = "#808080") +   #high completeness quality
  annotate("text", x=95, y=100, label="High", 
            colour="black", text=element_text(size=9, family="Arial")) +
  geom_segment(aes(x = 91, y = 96, xend = 100, yend = 96),
                  arrow = arrow(length = unit(0.25, "cm")), colour="black")+
  #-- LOW CONTAMINATION --
  annotate("text", x=110, y=0, label="Low", #angle=270,
            colour="black", text=element_text(size=9, family="Arial")) +
  geom_segment(aes(x = 105, y = 0, xend = 105, yend = 5),
                  arrow = arrow(length = unit(0.25, "cm")), colour="black")+
  #-- MED CONTAMINATION --
  annotate("text", x=110, y=8, label="Med", #angle=270,
            colour="black", text=element_text(size=9, family="Arial")) +
  geom_segment(aes(x = 105, y = 5, xend = 105, yend = 10),
                  arrow = arrow(length = unit(0.25, "cm")), colour="black")+
  geom_hline(yintercept = 5, linetype = "dashed", size = 1, color = "#808080") +    #med contamination quality = 5-10
  annotate("text", x=110, y=50, label="High", #angle=270,
            colour="black", text=element_text(size=9, family="Arial")) +
  geom_segment(aes(x = 105, y = 10, xend = 105, yend = 100),
                  arrow = arrow(length = unit(0.25, "cm")), colour="black")+
  geom_hline(yintercept = 10, linetype = "dashed", size = 1, color = "#808080") +   #high contamination quality = > 10%
  xlab("Completeness (%)") +
  ylab("Contamination (%)") +
  theme_Publication_legend_side()
  # guides(color=guide_legend(nrow=5)) 
  # labs(title="Completeness vs Contamination by Lineage")

# ----- PLOT WITH ONLY THE DATA WE ARE GOING TO ANALYSE -----
#This includes the medium-high quality MAGS at depth 200 and the cruises: 48, 72, 73. 74, and 75
# We are not analyzing Cruise 42, even though we have access to it because all other cruises came from August and cruise 42 came from february. We know that seasons can affect the compostion and we didn't have enough data points to generalize for other seasons than summer (August)

rpkm_dat %>%
  dplyr::filter(Contamination <= 10) %>%
  dplyr::filter(Completeness > 50) %>%
  ggplot(aes(x=Completeness, y=Contamination, colour=Class)) +
  geom_point(aes(size=g_rpkm)) +
  #added discrete scales ; palette from colorbrewer2 but added extra ("#969696") b/c 13 values
  # scale_color_manual(values = c('#a6cee3','#1f78b4',"#b2df8a","#33a02c","#fb9a99",
                                # "#e31a1c",'#fdbf6f',"#ff7f00","#cab2d6","#6a3d9a",
                                # "#ffff99","#b15928", "#969696")) + 
  #-- MED COMPLETENESS--
  geom_vline(xintercept = 50, linetype = "dashed", size = 1, color = "#808080") +   #med completeness quality
  annotate("text", x=70, y=12, label="Medium", 
            colour="black", text=element_text(size=9, family="Arial")) +
  geom_segment(aes(x = 51, y = 11, xend = 89, yend = 11),
                  arrow = arrow(length = unit(0.25, "cm")), colour="black")+
  geom_segment(aes(x = 89, y = 11, xend = 51, yend = 11),
                  arrow = arrow(length = unit(0.25, "cm")), colour="black")+
  #-- HIGH COMPLETENESS --
  geom_vline(xintercept = 90, linetype = "dashed", size = 1, color = "#808080") +   #high completeness quality
  annotate("text", x=95, y=12, label="High", 
            colour="black", text=element_text(size=9, family="Arial")) +
  geom_segment(aes(x = 91, y = 11, xend = 100, yend = 11),
                  arrow = arrow(length = unit(0.25, "cm")), colour="black")+
  geom_segment(aes(x = 100, y = 11, xend = 91, yend = 11),
                  arrow = arrow(length = unit(0.25, "cm")), colour="black")+
  #-- LOW CONTAMINATION --
  annotate("text", x=104, y=2.5, label="Low", #angle=270,
            colour="black", text=element_text(size=9, family="Arial")) +
  geom_segment(aes(x = 101, y = 0, xend = 101, yend = 4.8),
                  arrow = arrow(length = unit(0.25, "cm")), colour="black")+
  geom_segment(aes(x = 101, y = 4.8, xend = 101, yend = 0),
                  arrow = arrow(length = unit(0.25, "cm")), colour="black")+
  #-- MED CONTAMINATION --
  annotate("text", x=104, y=8, label="Med", #angle=270,
            colour="black", text=element_text(size=9, family="Arial")) +
  geom_segment(aes(x = 101, y = 5.2, xend = 101, yend = 10),
                  arrow = arrow(length = unit(0.25, "cm")), colour="black")+
  geom_segment(aes(x = 101, y = 10, xend = 101, yend = 5.2),
                  arrow = arrow(length = unit(0.25, "cm")), colour="black")+
  geom_hline(yintercept = 5, linetype = "dashed", size = 1, color = "#808080") +    #med contamination quality = 5-10
  xlab("Completeness (%)") +
  ylab("Contamination (%)") +
  theme_Publication_legend_side() 
  # guides(color=guide_legend(nrow=5)) 

```

# Bubble Plot
Next I will create RPKM bubble-plot of each Sulphur-cycling gene versus taxonomy
```{r}
#TODO: add category stripe of phylum

#Data frame of ko values and their corresponding step in the pathway (found manually from pathview output)
ko_to_step <- read_csv("KO-StepName-Mapping.csv")

#read in the pathview output from sulfur pathway grouped by class (generated in pathview_by_class Rmd file)
pv_gene_dat <- readRDS("pv_out_mean.rds")

#Filter by the ko values that were found to be involved in the pathways steps we are interested in
pv_steps_dat <- dplyr::inner_join(pv_gene_dat, ko_to_step)
pv_steps_dat <- pv_steps_dat[, -grep(".col$", colnames(pv_steps_dat))]
names(pv_steps_dat)[names(pv_steps_dat) == 'c__'] <- 'c__Unknown'

#wide to long format
pv_steps_dat <- pv_steps_dat %>%
  tidyr::gather(key = Class, value = rpkm, -c(kegg.names, labels, KO, all.mapped, type, width, height, StepName, x, y))

#Create a bubble plot
pv_steps_dat %>%
  mutate(Class = gsub("c__", "", Class)) %>%
  ggplot(aes(x=StepName, y=Class, size = rpkm, color = rpkm)) + 
  geom_point() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  # labs(title = "Class vs Steps of Interest in Sulfur Metabolic Pathway") +
  scale_color_gradientn(limits = c(0,5),
                        colours = c("#b10026",  "#e31a1c","#fc4e2a", "#fd8d3c", "#feb24c", "#fed976"),
                        breaks = c(0,1,2,3,4,5),
                        guide = guide_colorbar(barwidth = 1, barheight = 5)) +
  theme_Publication_legend_side() +
  scale_size(range=c(2, 6)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.text = element_text(margin = margin(t = 10)),
        legend.spacing.y = unit(0.5, 'cm')) 


```

# Bubble Plot from gamma proteobacteria
Next I will create RPKM bubble-plot of each Nitrogen/Sulphur-cycling gene versus taxonomy
```{r}
#TODO: add category stripe of phylum

#Data frame of ko values and their corresponding step in the pathway (found manually from pathview output)
ko_to_step <- read_csv("KO-StepName-Mapping.csv")

#read in the pathview output from sulfur pathway grouped by class (generated in pathview_by_class Rmd file)
pv_gene_dat <- readRDS("gamma_pv_out.rds")

#Filter by the ko values that were found to be involved in the pathways steps we are interested in
pv_steps_dat <- dplyr::inner_join(pv_gene_dat, ko_to_step)
pv_steps_dat <- pv_steps_dat[, -grep(".col$", colnames(pv_steps_dat))]
names(pv_steps_dat)[names(pv_steps_dat) == 'o__'] <- 'o__Unknown'

#wide to long format
pv_steps_dat <- pv_steps_dat %>%
  tidyr::gather(key = Order, value = rpkm, -c(kegg.names, labels, KO, all.mapped, type, width, height, StepName, x, y))

#Create a bubble plot
pv_steps_dat %>%
  mutate(Order = gsub("o__", "", Order)) %>%
  ggplot(aes(x=StepName, y=Order)) + 
  geom_point(aes(size = rpkm, color=rpkm)) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  # labs(title = "Order of Gammaproteobacteria vs Steps in Sulfur Metabolic Pathway") +
  scale_color_gradientn(limits = c(0,5),
                        colours = c("#b10026",  "#e31a1c","#fc4e2a", "#fd8d3c", "#feb24c", "#fed976"),
                        breaks = c(0,1,2,3,4,5),
                        guide = guide_colorbar(barwidth = 1, barheight = 5)) +
  theme_Publication_legend_side() +
  scale_size(range=c(2, 6)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.text = element_text(margin = margin(t = 10)),
        legend.spacing.y = unit(0.5, 'cm')) 


```



# Geochemical Gradients
```{r}
# Load raw data
raw_dat <- readr::read_csv("Saanich_Data.csv")

#Clean dat
dat <- 
  raw_dat %>%
  dplyr::select(Cruise, Date, Depth, Temperature,
         WS_O2, WS_NO3, WS_H2S) %>%
  dplyr::filter(!is.na(WS_O2)) %>%
  dplyr::rename(O2_uM=WS_O2, NO3_uM=WS_NO3, H2S_uM=WS_H2S) %>%
  dplyr::mutate(Depth_m=Depth*1000) %>%
  dplyr::filter(Cruise %in% c(48, 72, 73, 74, 75)) #Only take from the relevant cruises

###############
# -- First let's look at all of the depths, so we can get a high level picture of the concentration changes with depth --
#want to create a plot with all 3 variables as x axis and depth as y with shape as type
#can manipulate data frame to do so
dat %>%
  dplyr::select(Depth_m, H2S_uM, NO3_uM, O2_uM) %>%
  gather(key = "Chemical", value = "Concentration", -Depth_m) %>%
  ggplot() +
  geom_point(aes(x=Concentration, y=Depth_m, colour=Chemical)) +
  scale_y_reverse() +
  expand_limits(y=0) +
  theme_Publication() +
  scale_color_manual(values=c("#e41a1c", "#377eb8", "#4daf4a"))

# -- Now let's facet to get a better look --
dat %>%
  dplyr::select(Depth_m, H2S_uM, NO3_uM, O2_uM) %>%
  gather(key = "Chemical", value = "Concentration", -Depth_m) %>%
  ggplot() +
  geom_point(aes(x=Concentration, y=Depth_m, colour=Chemical)) +
  scale_y_reverse() +
  expand_limits(y=0) +
  theme_Publication_facets() +
  scale_color_manual(values=c("#e41a1c", "#377eb8", "#4daf4a")) +
  facet_wrap(facets = vars(Chemical))


###############
# -- Next let's zoom into depth 200 --

#Normal
dat %>%
  dplyr::select(Depth_m, H2S_uM, NO3_uM, O2_uM) %>%
  dplyr::filter(Depth_m > 150) %>%
  dplyr::filter(Depth_m <= 200) %>%
  gather(key = "Chemical", value = "Concentration", -Depth_m) %>%
  ggplot() +
  geom_point(aes(x=Concentration, y=Depth_m, colour=Chemical)) +
  scale_y_reverse() +
  expand_limits(y=150) +
  theme_Publication() +
  scale_color_manual(values=c("#e41a1c", "#377eb8", "#4daf4a"))

#facet
dat %>%
  dplyr::select(Depth_m, H2S_uM, NO3_uM, O2_uM) %>%
  dplyr::filter(Depth_m > 150) %>%
  dplyr::filter(Depth_m <= 200) %>%
  gather(key = "Chemical", value = "Concentration", -Depth_m) %>%
  ggplot() +
  geom_point(aes(x=Concentration, y=Depth_m, colour=Chemical)) +
  scale_y_reverse() +
  expand_limits(y=150) +
  theme_Publication_facets() +
  scale_color_manual(values=c("#e41a1c", "#377eb8", "#4daf4a")) +
  facet_wrap(facets = vars(Chemical))
  

```

For this next plot, we will determine the relative abundance of these MAGs across the different cruises and time points. We don't want to use the mean RPKM value across the cruises so we are going to create a new dataframe for this plot. We'll begin exactly as before, but we're going to filter the MAGs down to something that isn't crowded and visually overwhelming. 
I selected all the medium- and high-quality Proteobacteria, then sorted by taxonomic Order. This will make `ggplot` plot all the MAGs that are taxonomically related together, rather than by their arbitrary numerical identifiers. We're also discarding all the low-quality MAGs since they are full of sequences from who-knows-what. Total Frankenbacteria!


```{r cruise_bubbles}
rpkm_dat <- left_join(metag_rpkm, checkm_dat, by="mag") %>% 
  left_join(gtdb_dat, by="mag") %>% 
  filter(Completeness > 50 & Contamination < 10) %>% # good quality MAGs
  filter(Phylum == "p__Proteobacteria") %>% # Proteobacteria only; this can easily be changed
  mutate(mag = reorder(mag, Order, sort)) # sort by their taxonomic Order so everything shows up together
ggplot(rpkm_dat, aes(x=Sample, y=mag, col=Order)) +
  geom_point(aes(size=g_rpkm)) +
  theme_Publication_legend_side() +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1))
```
