---
title: "Pathview Aggregated by Order"
author: "Shannah"
date: "November 25, 2018"
output: html_document
---

```{r setup, warning=FALSE, message=FALSE}
library(tidyr)
library(dplyr)
library(pathview)
library(RColorBrewer)
library(knitr)
```

Now we are going to read our tabular files. First is KO annotation tables (query.ko.txt) then the rpkm file. I'm using rename to assign names to each of the columns as these do not have headers.

```{r, read, warning=FALSE, message=FALSE}

#Read in KO Annotation tables
ko <- read.table("SaanichInlet_MAGs_ORFs_ko.cleaned.txt") %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(ko = V2)

#Read in all rpkm files from each cruise
# rpkm_42 <- read.table("SI042_200m.RPKM.csv", header=FALSE, sep=',') %>% 
#   dplyr::rename(orf = V1) %>% 
#   dplyr::rename(rpkm = V2)
rpkm_48 <- read.table("SI048_200m.RPKM.csv", header=FALSE, sep=',') %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(rpkm = V2)
rpkm_72 <- read.table("SI048_200m.RPKM.csv", header=FALSE, sep=',') %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(rpkm = V2)
rpkm_73 <- read.table("SI073_200m.RPKM.csv", header=FALSE, sep=',') %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(rpkm = V2)
rpkm_74 <- read.table("SI074_200m.RPKM.csv", header=FALSE, sep=',') %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(rpkm = V2)
rpkm_75 <- read.table("SI075_200m.RPKM.csv", header=FALSE, sep=',') %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(rpkm = V2)

#Read in prokka MAG map
prokka_mag_map <- read.table("Prokka_MAG_map_basename.csv", header=F, sep=',') %>% 
  dplyr::rename(prokka_id = V1) %>% 
  dplyr::rename(mag = V2)

#Read in gtdbtk files
arc_class <- read.table("gtdbtk.ar122.classification_pplacer.tsv", sep="\t")
bac_class <- read.table("gtdbtk.bac120.classification_pplacer.tsv", sep="\t")

#Combine archaea and bacteria gtdb files
gtdb_dat <- rbind(arc_class, bac_class) %>% 
  dplyr::rename(mag = V1) %>% 
  separate(V2, sep=';', into=c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"))

#checkm data
checkm_dat <- read.table("MetaBAT2_SaanichInlet_200m_min1500_checkM_stdout.tsv",
                         header=TRUE,
                         sep="\t",
                         comment.char = '') %>% 
  dplyr::rename(mag = Bin.Id) %>% 
  dplyr::select(mag, Completeness, Contamination)

# Due to a bug in the renaming script we have to rename the bins. Its a bit hacky but works using tidyverse functions
metag_rpkm <- read.table("SaanichInlet_200m_binned.rpkm.csv", header=T, sep=',') %>% 
  mutate(Sequence = gsub('m_', 'm.', Sequence)) %>% 
  mutate(Sequence = gsub('Inlet_', 'Inlet.', Sequence)) %>% 
  separate(col=Sequence, into=c("mag", "contig"), sep='_', extra="merge") %>% 
  group_by(Sample, mag) %>% 
  summarise(g_rpkm = mean(RPKM)) %>% 
  mutate(mag = gsub('Inlet.', 'Inlet_', mag))
```

We can also determine the number of Phyla present in our bins, and how many MAGs are representing each. Of course, you can do this for all the taxonomic ranks. I'm also going to remove the Genus and Species fields since these are probably not relevant to your analysis (if so, then skip that step) and not very trustworthy given the whole binning procedure.
```{r diversity}
gtdb_dat %>% 
  group_by(Order) %>% 
  summarise(count = n_distinct(mag)) %>% 
  kable()
gtdb_dat <- dplyr::select(gtdb_dat, mag, Kingdom, Phylum, Class, Order, Family)
```

Next steps are to join these two tables and some basic transformations to make things a bit easier for `pathview`. 
I've separated, or split, the `orf` value of each row into two new variables: `mag` and `orf_id` corresponding to the character string before and after the underscore in `orf`. This makes it easier to `group_by` MAGs and will be necessary for joining other tables (such as checkM, gtdbtk, etc.) into one dataframe (to rule them all).

The `summarise` code is summing all RPKM values assigned to a KO number for each MAG. This is useful to prevent multiple rows in an eventual matrix for pathview for each copy found. Or accidentally dropping those data if we're not careful. Anyway, we can freely sum RPKM values and that is what is easiest here.

__NOTE__: If your are dealing with RPKM values from multiple cruises (in this example I am only dealing with RPKM from SI042) you will also need to group by a `cruise` variable so these are not summed. Or maybe you want them to be if you are not interested in the time/season/cruise variable. 
If you are interested in visualizing the variability in transcription of a single MAG across the cruises you may also want to `filter` for your MAG of interest then group by `ko` and `cruise`. It all depends on what question you want to answer so be mindful here!

```{r, arrange, warning=FALSE, message=FALSE}
ko_rpkm_48 <- left_join(ko, rpkm_48, by="orf") %>% 
  separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF
  left_join(prokka_mag_map, by="prokka_id") %>% 
  left_join(gtdb_dat, by="mag") %>%
  left_join(checkm_dat, by="mag") %>%
  left_join(metag_rpkm, by="mag")
ko_rpkm_72 <- left_join(ko, rpkm_72, by="orf") %>% 
  separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF
  left_join(prokka_mag_map, by="prokka_id") %>% 
  left_join(gtdb_dat, by="mag")%>%
  left_join(checkm_dat, by="mag") %>%
  left_join(metag_rpkm, by="mag")
ko_rpkm_73 <- left_join(ko, rpkm_73, by="orf") %>% 
  separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF
  left_join(prokka_mag_map, by="prokka_id") %>% 
  left_join(gtdb_dat, by="mag")%>%
  left_join(checkm_dat, by="mag") %>%
  left_join(metag_rpkm, by="mag")
ko_rpkm_74 <- left_join(ko, rpkm_74, by="orf") %>% 
  separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF
  left_join(prokka_mag_map, by="prokka_id") %>% 
  left_join(gtdb_dat, by="mag")%>%
  left_join(checkm_dat, by="mag") %>%
  left_join(metag_rpkm, by="mag")
ko_rpkm_75 <- left_join(ko, rpkm_75, by="orf") %>% 
  separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF
  left_join(prokka_mag_map, by="prokka_id") %>% 
  left_join(gtdb_dat, by="mag")%>%
  left_join(checkm_dat, by="mag") %>%
  left_join(metag_rpkm, by="mag")

rpkm_dat <- rbind(ko_rpkm_48, ko_rpkm_73, ko_rpkm_74, ko_rpkm_75)

head(rpkm_dat) %>% kable()


#Commenting out connors tutorial code
# rpkm_dat <- left_join(ko, metat_rpkm, by="orf") %>%
#   separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF number for joining
#   left_join(prokka_mag_map, by="prokka_id") %>% 
#   left_join(gtdb_dat, by="mag") %>% 
#   left_join(checkm_dat, by="mag")
#
# If you also wanted to add the RPKM abundance values from the metagenome:
  # left_join(metag_rpkm, by="mag")
```

Depending on your research question, you may want to subset your data by just the MAGs with the greatest completeness and least contamination, or by a taxonomy. Or, if you are wondering what the metabolic potential is of the entire community (well, at least the portion that was binned) you may want to aggregate the KO numbers by a taxonomic rank. 

__NOTE__: some sort of aggregation or subsetting is probably essential with `pathview` since it has an upper limit of the number of samples that can be visualized... and by sample I mean MAG, cruise, or taxon. Really, this comes down to the number of columns in the final matrix you pass off to `pathview`. However, I recently found that updating your R version to 3.5.1 and installing the latest version of `pathview` increases this number to some as of yet unknown limit!

```{r subsetting, warning=FALSE, message=FALSE}
# Subset by taxon (COMMENTING OUT BECASUE WE DON'T WANT TO DO THIS)
# ko_rpkm <- rpkm_dat %>%
#   filter(Phylum %in% c("p__Proteobacteria", "p__Nanoarchaeota", "p__Thermoplasmatota")) %>%
#   group_by(mag, ko) %>% 
#   summarise(t_rpkm = sum(rpkm)) %>% 
#   spread(key = mag, value = t_rpkm)

# Subset by completeness and contamination (COMMENTING OUT BECAUSE WE DON'T WANT TO DO THIS)
# ko_rpkm <- rpkm_dat %>% 
#   filter(Completeness >= 90 & Contamination < 5) %>% 
#   group_by(mag, ko) %>% 
#   summarise(t_rpkm = sum(rpkm)) %>% 
#   spread(key = mag, value = t_rpkm)

# Aggregate by order, and summing RPKM of each KO number. [TODO: ? You could use mean() instead.]
lo_orders <- unique(rpkm_dat$Order)

lo_rpkm_mat <- lapply(lo_orders, function(order) {
  t_rpkm <- rpkm_dat %>%
    dplyr::filter(Order == order) %>%
    dplyr::group_by(Order, ko) %>% 
    dplyr::summarise(t_rpkm = mean(rpkm)) %>% 
    spread(key = Order, value = t_rpkm)
    
  pv_mat <- dplyr::select(t_rpkm, -ko)
  rownames(pv_mat) <- t_rpkm$ko
  return(pv_mat)
})


#Name each element in the list according to the order (for organization)
names(lo_rpkm_mat) <- lo_orders

#For a single pathview output that is grouped by order
#We want to dig deeper into the prot
t_rpkm <- rpkm_dat %>%
  dplyr::filter(Class == "c__Gammaproteobacteria") %>%
  dplyr::group_by(Order, ko) %>% 
  dplyr::summarise(t_rpkm = mean(rpkm, na.rm=TRUE)) %>% 
  spread(key = Order, value = t_rpkm)
  
pv_mat <- dplyr::select(t_rpkm, -ko)
rownames(pv_mat) <- t_rpkm$ko

```

We will be viewing the sulfur metabolic pathway

color scale using [colorbrewer2](http://colorbrewer2.org/#type=sequential&scheme=BuGn&n=3). 

```{r, pathviewing, warning=FALSE, message=FALSE}
lo_orders <- unique(rpkm_dat$Order)

# Sulfur metabolism pathview for each order
# lo_pv.out_order <- lapply(lo_orders, function(order) {
  # setwd(paste0("C:/Users/Shannah/OneDrive/Uni-Year-5/MICB_405/Project2/pathview_output/", order, "/"))
  # pathview(gene.data = lo_rpkm_mat[[]],
  #          limit = list(gene = c(0,10)),
  #          low = list(gene = "#91bfdb"),
  #          mid = list(gene = "#ffffbf"),
  #          high = list(gene = "#fc8d59"),
  #          species = "ko",
  #          pathway.id="00920",
  #          kegg.dir=".")
  
# })

#Sulfur metabolism pathview grouped by order
pv.out <- pathview(gene.data = pv_mat,
                   limit = list(gene = c(0,10)),
                   low = list(gene = "#91bfdb"),
                   mid = list(gene = "#ffffbf"),
                   high = list(gene = "#fc8d59"),
                   species = "ko",
                   pathway.id="00920",
                   kegg.dir = ".")
saveRDS(pv.out$plot.data.gene, "gamma_pv_out.rds")

```



