---
title: "Pathview"
author: "Team 17"
date: "November 23, 2018"
output: html_document
---

This tutorial is meant to take your KEGG Ortholog (KO) annotation tables from KAAS, and one of the rpkm files (.csv format) and view a pathway or metabolic map with these data layered on.

Before you begin loading, transforming, and playing with data we will need to load the following libraries at the very least:

```{r setup, warning=FALSE, message=FALSE}
library(tidyr)
library(dplyr)
library(pathview)
```

Now we are going to read our tabular files. First is KO annotation tables (query.ko.txt) then the rpkm file. I'm using rename to assign names to each of the columns as these do not have headers.

```{r, read, warning=FALSE, message=FALSE}

#Read in KO Annotation tables
ko <- read.table("SaanichInlet_MAGs_ORFs_ko.cleaned.txt") %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(ko = V2)

#Read in all rpkm files from each cruise
# rpkm_42 <- read.table("SI042_200m.RPKM.csv", header=FALSE, sep=',') %>% 
#   dplyr::rename(orf = V1) %>% 
#   dplyr::rename(rpkm = V2)
rpkm_48 <- read.table("SI048_200m.RPKM.csv", header=FALSE, sep=',') %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(rpkm = V2)
rpkm_72 <- read.table("SI048_200m.RPKM.csv", header=FALSE, sep=',') %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(rpkm = V2)
rpkm_73 <- read.table("SI073_200m.RPKM.csv", header=FALSE, sep=',') %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(rpkm = V2)
rpkm_74 <- read.table("SI074_200m.RPKM.csv", header=FALSE, sep=',') %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(rpkm = V2)
rpkm_75 <- read.table("SI075_200m.RPKM.csv", header=FALSE, sep=',') %>% 
  dplyr::rename(orf = V1) %>% 
  dplyr::rename(rpkm = V2)

#Read in prokka MAG map
prokka_mag_map <- read.table("Prokka_MAG_map_basename.csv", header=F, sep=',') %>% 
  dplyr::rename(prokka_id = V1) %>% 
  dplyr::rename(mag = V2)

#Read in gtdbtk files
arc_class <- read.table("gtdbtk.ar122.classification_pplacer.tsv", sep="\t")
bac_class <- read.table("gtdbtk.bac120.classification_pplacer.tsv", sep="\t")

#Combine archaea and bacteria gtdb files
gtdb_dat <- rbind(arc_class, bac_class) %>% 
  dplyr::rename(mag = V1) %>% 
  separate(V2, sep=';', into=c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"))
```

Next steps are to join these two tables and some basic transformations to make things a bit easier for `pathview`. 
I've separated, or split, the `orf` value of each row into two new variables: `mag` and `orf_id` corresponding to the character string before and after the underscore in `orf`. This makes it easier to `group_by` MAGs and will be necessary for joining other tables (such as checkM, gtdbtk, etc.) into one dataframe (to rule them all).

The `summarise` code is summing all RPKM values assigned to a KO number for each MAG. This is useful to prevent multiple rows in an eventual matrix for pathview for each copy found. Or accidentally dropping those data if we're not careful. Anyway, we can freely sum RPKM values and that is what is easiest here.

__NOTE__: If your are dealing with RPKM values from multiple cruises (in this example I am only dealing with RPKM from SI042) you will also need to group by a `cruise` variable so these are not summed. Or maybe you want them to be if you are not interested in the time/season/cruise variable. 
If you are interested in visualizing the variability in transcription of a single MAG across the cruises you may also want to `filter` for your MAG of interest then group by `ko` and `cruise`. It all depends on what question you want to answer so be mindful here!

```{r, arrange, warning=FALSE, message=FALSE}

#combine all cruises and ko values
#---
#February
# ko_rpkm_42 <- left_join(ko, rpkm_42, by="orf") %>% 
#   separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF
#   left_join(prokka_mag_map, by="prokka_id") %>% 
#   left_join(gtdb_dat, by="mag")
#---
ko_rpkm_48 <- left_join(ko, rpkm_48, by="orf") %>% 
  separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF
  left_join(prokka_mag_map, by="prokka_id") %>% 
  left_join(gtdb_dat, by="mag")
ko_rpkm_72 <- left_join(ko, rpkm_72, by="orf") %>% 
  separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF
  left_join(prokka_mag_map, by="prokka_id") %>% 
  left_join(gtdb_dat, by="mag")
ko_rpkm_73 <- left_join(ko, rpkm_73, by="orf") %>% 
  separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF
  left_join(prokka_mag_map, by="prokka_id") %>% 
  left_join(gtdb_dat, by="mag")
ko_rpkm_74 <- left_join(ko, rpkm_74, by="orf") %>% 
  separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF
  left_join(prokka_mag_map, by="prokka_id") %>% 
  left_join(gtdb_dat, by="mag")
ko_rpkm_75 <- left_join(ko, rpkm_75, by="orf") %>% 
  separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF
  left_join(prokka_mag_map, by="prokka_id") %>% 
  left_join(gtdb_dat, by="mag")

#Merge all cruises except for cruise 42 (from february)
#All cruises are from August except for Cruise 42 and b/c seasons can dramatically affect the water column AND we don't have enough data points to compare across seasons (only one from February) we are removing the February rpkm value.
ko_rpkm <- rbind(ko_rpkm_48, ko_rpkm_73, ko_rpkm_74, ko_rpkm_75)

#Average values for a given orf for a given mag

#Commenting out code from connor without individual cruises
# ko_rpkm <- left_join(ko, rpkm, by="orf") %>% 
#   separate(orf, into=c("prokka_id", "orf_id")) %>% # Split the Prokka ORF names into MAG identifier and ORF number for joining
#   left_join(prokka_mag_map, by="prokka_id") %>% 
#   left_join(gtdb_dat, by="mag")

#Get all phylum
lo_phylum <- unique(ko_rpkm$Phylum)

#Get a list of rpkm dataframes filtering by phylum and then create a matrix 
#We are summing all cruises together because we do not care about the cruise/season/time variable for our question
lo_rpkm_mat <- lapply(lo_phylum, function(phylum) {
  t_rpkm <- ko_rpkm %>% 
    dplyr::filter(Phylum == phylum) %>% 
    dplyr::group_by(mag, ko) %>% 
    dplyr::summarise(total = sum(rpkm)) %>% 
    spread(key = mag, value = total)
  
  pv_mat <- dplyr::select(t_rpkm, -ko)
  rownames(pv_mat) <- t_rpkm$ko
  return(pv_mat)
})

#Name each element in the list according to the phylum (for organization)
names(lo_rpkm_mat) <- lo_phylum

#commenting out code from connor without loop
# test_rpkm <- ko_rpkm %>%
#   dplyr::filter(Phylum == "p__Marinisomatota") %>%
#   dplyr::group_by(mag, ko) %>%
#   dplyr::summarise(total = sum(rpkm)) %>%
#   spread(key = mag, value = total)
# test_mat <- dplyr::select(test_rpkm, -ko)
# rownames(test_mat) <- test_rpkm$ko

```

As an example, we can view the nitrogen metabolism capabilities of our MAGs. To view a different pathway or metabolism the `pathway.id` parameter will need to be changed. Searching for your pathway of interest via the KEGG browser is likely the easiest way to find these IDs.

We can also view the dataframe that is generated by `pathview`. Unfortunately it is not that interesting or useful.

```{r, pathviewing, warning=FALSE, message=FALSE}

#Create the pathview images for each pathway () and each phylum

lo_pathway_ids <- c(sulfur="00920", carbon="01200", nitrogen="00910", carbon_fix="00720") #TODO: double check carbon


#Ideally we would like to make a loop for each of the phylums but pathview output is based on the pathway.id, and there is no way that we could find to specify the output file name otherwise. So if we were to create a loop it would override the file each time it looped. kegg.dir only writes the png and xml but not the multipng. So our work around was to create each of the files one at a time and manually rename them.
lo_Chloroflexota_pv.out <- lapply(lo_pathway_ids, function(pathway_id) {
  pathview(gene.data = lo_rpkm_mat[["p__Chloroflexota"]],
           species = "ko",
           pathway.id=pathway_id)
})

lo_Patescibacteria_pv.out <- lapply(lo_pathway_ids, function(pathway_id) {
  pathview(gene.data = lo_rpkm_mat[["p__Patescibacteria"]],
           species = "ko",
           pathway.id=pathway_id)
})

lo_Nanoarchaeota_pv.out <- lapply(lo_pathway_ids, function(pathway_id) {
  pathview(gene.data = lo_rpkm_mat[["p__Nanoarchaeota"]],
           species = "ko",
           pathway.id=pathway_id)
})

lo_Elusimicrobiota_pv.out <- lapply(lo_pathway_ids, function(pathway_id) {
  pathview(gene.data = lo_rpkm_mat[["p__Elusimicrobiota"]],
           species = "ko",
           pathway.id=pathway_id)
})

lo_Proteobacteria_pv.out <- lapply(lo_pathway_ids, function(pathway_id) {
  pathview(gene.data = lo_rpkm_mat[["p__Proteobacteria"]],
           species = "ko",
           pathway.id=pathway_id)
})

lo_SAR324_pv.out <- lapply(lo_pathway_ids, function(pathway_id) {
  pathview(gene.data = lo_rpkm_mat[["p__SAR324"]],
           species = "ko",
           pathway.id=pathway_id)
})

lo_Micrarchaeota_pv.out <- lapply(lo_pathway_ids, function(pathway_id) {
  pathview(gene.data = lo_rpkm_mat[["p__Micrarchaeota"]],
           species = "ko",
           pathway.id=pathway_id)
})

lo_Marinisomatota_pv.out <- lapply(lo_pathway_ids, function(pathway_id) {
  pathview(gene.data = lo_rpkm_mat[["p__Marinisomatota"]],
           species = "ko",
           pathway.id=pathway_id)
})

lo_Crenarchaeota_pv.out <- lapply(lo_pathway_ids, function(pathway_id) {
  pathview(gene.data = lo_rpkm_mat[["p__Crenarchaeota"]],
           species = "ko",
           pathway.id=pathway_id)
})

lo_Bacteroidota_pv.out <- lapply(lo_pathway_ids, function(pathway_id) {
  pathview(gene.data = lo_rpkm_mat[["p__Bacteroidota"]],
           species = "ko",
           pathway.id=pathway_id)
})

lo_Dependentiae_pv.out <- lapply(lo_pathway_ids, function(pathway_id) {
  pathview(gene.data = lo_rpkm_mat[["p__Dependentiae"]],
           species = "ko",
           pathway.id=pathway_id)
})

lo_Gemmatimonadota_pv.out <- lapply(lo_pathway_ids, function(pathway_id) {
  pathview(gene.data = lo_rpkm_mat[["p__Gemmatimonadota"]],
           species = "ko",
           pathway.id=pathway_id)
})


# Commenting out from connor's tutorial
# # Nitrogen metabolism
# pv.out <- pathview(gene.data = pv_mat,
#                    species = "ko",
#                    pathway.id="00910",
#                    kegg.dir = "~/Desktop/MICB405_TAship/MICB405-Metagenomics/2018/")
```


Here is a file that pathview automatically writes to your `kegg.dir` directory.

!["An example nitrogen metabolism wiring diagram derived from Saanich Inlet MAGs. There are no negative RPKM values and therefore the lowest value is 0. Vertical red bars indicate some MAGs are actively transcirbing genes involved in disimilatory nitrate reduction, denitrification and nitrification."](/home/connor/Desktop/MICB405_TAship/MICB405-Metagenomics/2018/ko00910.pathview.multi.png)
